etcd-recovery/etcdrecovery.go

```diff
@@ -0,0 +1,391 @@
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "etcd",
+			Namespace: hcpNS,
+		},
+	}
+
+	if err := k.Get(ctx, crclient.ObjectKeyFromObject(etcdSTS), etcdSTS); err != nil {
+		return fmt.Errorf("failed to get etcd statefulset: %w", err), nil, nil
+	}
+
+	if etcdSTS.Status.CurrentRevision != etcdSTS.Status.UpdateRevision {
+		return fmt.Errorf("etcd statefulset is updating, skipping recovery procedure"), nil, nil
+	}
+
+	return nil, failingEndpoint, &failingPods[0]
+}
+
+// etcdValidate expects a list of endpointStatuses and validates the status of the etcd cluster.
+// At this point the expectation is to have at least 3 endpoints and 1 of them in error state.
+func etcdValidate(epStatuses []endpointStatus) (*endpointStatus, error) {
+	var errorCount int
+	var failingEndpoint endpointStatus
+
+	if len(epStatuses) != 3 {
+		return nil, fmt.Errorf("expected 3 etcd endpoints, got %d", len(epStatuses))
+	}
+
+	for _, epStatus := range epStatuses {
+		if len(epStatus.errors) > 0 {
+			errorCount++
+			failingEndpoint = epStatus
+		}
+	}
+
+	if errorCount <= 0 {
+		fmt.Println("etcd cluster is healthy")
+		return nil, nil
+	}
+
+	if errorCount > 1 {
+		return nil, fmt.Errorf("expected 1 etcd endpoint in error state, got %d", errorCount)
+	}
+
+	// three different hashes are not expected
+	if epStatuses[0].hash != epStatuses[1].hash && epStatuses[1].hash != epStatuses[2].hash {
+		return nil, fmt.Errorf("etcd cluster hash mismatch: %v", epStatuses)
+	}
+
+	return &failingEndpoint, nil
+}
+
+func recoverEtcd(ctx context.Context, failingEndpoint *endpointStatus, failingPod corev1.Pod, k crclient.Client) error {
+	fmt.Printf("Recovering etcd cluster, endpoint: %s name: %s\n", failingEndpoint.endpoint, failingEndpoint.name)
+	hcpNS := os.Getenv("NAMESPACE")
+
+	// Checks finished starting recovery procedure
+	etcdPV := &corev1.PersistentVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("data-%s", failingPod.Name),
+			Namespace: hcpNS,
+		},
+	}
+
+	if err := k.Get(ctx, crclient.ObjectKeyFromObject(etcdPV), etcdPV); err != nil {
+		return fmt.Errorf("failed to get etcd persistent volume: %w", err)
+	}
+
+	if etcdPV.Spec.ClaimRef == nil {
+		return fmt.Errorf("etcd persistent volume has no claim reference: %v", etcdPV)
+	}
+
+	etcdPVC := &corev1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      etcdPV.Spec.ClaimRef.Name,
+			Namespace: hcpNS,
+		},
+	}
+
+	if err := k.Get(ctx, crclient.ObjectKeyFromObject(etcdPVC), etcdPVC); err != nil {
+		return fmt.Errorf("failed to get etcd persistent volume claim: %w", err)
+	}
+
+	if err := k.Delete(ctx, etcdPVC); err != nil {
+		return fmt.Errorf("failed to delete etcd persistent volume claim: %w", err)
+	}
+
+	if err := k.Delete(ctx, etcdPV); err != nil {
+		return fmt.Errorf("failed to delete etcd persistent volume: %w", err)
+	}
+
+	if err := k.Delete(ctx, &failingPod); err != nil {
+		return fmt.Errorf("failed to delete failing etcd pod: %w", err)
+	}
```

Comment:
@jparrill as an FYI, we've seen cases where these steps are not enough to recover the etcd member of an HCP stuck in installing. A full re-creation of the statefulset/PVs for all 3 etcd /PVCs for all 3 etcd was needed.

Reply:
The problem is that we have no good way of recovering from a missing member. The init container that re-adds the member only does it if the member already exists.  The new recovery code adds the member before deleting pod/pvc and that causes the recovery to succeed.